gbn/timeout_manager.go

```diff
@@ -62,11 +73,76 @@ func NewTimeOutManager(isServer bool) *TimeoutManager {
 		isServer:         isServer,
 		resendTimeout:    defaultResendTimeout,
 		handshakeTimeout: defaultHandshakeTimeout,
+		useStaticTimeout: false,
+		resendMultiplier: defaultResendMultiplier,
 		recvTimeout:      DefaultRecvTimeout,
 		sendTimeout:      DefaultSendTimeout,
 	}
 }
 
+// Sent is should be called when a message is sent by the connection.
+func (m *TimeoutManager) Sent(msg Message) {
+	//nolint:godox
+	// TODO(viktor): In the future, we may want to use this to keep track of
+	// the time it took for the other party to respond to other types of
+	// messages than the handshake, and dynamically keep updating the resend
+	// timeout to ensure that it reflects the current response time.
+	switch msg.(type) { //nolint:gocritic
+	case *PacketSYN:
+		// Note that we may send multiple SYN messages before receiving
+		// a response. Therefore, this field might be updated multiple
+		// times.
+		m.latestSentSYNTime = time.Now()
+	}
+}
+
+// Received is should be called when a message is received by the connection.
+func (m *TimeoutManager) Received(msg Message) {
+	//nolint:godox
+	// TODO(viktor): In the future, we may want to use this to keep track of
+	// the time it took for the other party to respond to other types of
+	// messages than the handshake, and dynamically keep updating the resend
+	// timeout to ensure that it reflects the current response time.
+	switch msg.(type) {
+	case *PacketSYN:
+		if !m.isServer {
+			m.updateResendTimeout(time.Since(m.latestSentSYNTime))
+		}
+
+	case *PacketSYNACK:
+		if m.isServer {
+			m.updateResendTimeout(time.Since(m.latestSentSYNTime))
+		}
+	}
+}
+
+// updateResendTimeout updates the resend timeout based on the given response
+// time. The resend timeout will be only be updated if the given response time
+// is greater than the default resend timeout, after being multiplied by the
+// resendMultiplier.
+// If a static timeout has been manually set, then this function will do be a
+// no-op.
+func (m *TimeoutManager) updateResendTimeout(responseTime time.Duration) {
```

Comment:
It currently could, but since we're moving to updating the resend timeout based on more messages then just the SYN & SYNACK messages I think it makes sense to keep this as an arg you pass to function, as the way you calculate how long the response took will vary based on different message types.